### ISSUE TITLE
LoRA merging doesn't work 

### ISSUE THREAD

**Issue by NicoZenith** [chuangchuangtan/LLaVA-NeXT-Image-Llama3-Lora#1]
Hi! Thank you again for this repo. The fine-tuning with llama3 works. However, when I try to merge with the obtained LoRA weights, using the `merge_lora_weights.py` script, and I compare the weights between the base model and the new merged model, I see changes in the vision encoder (which is surprising), and not in the bridger and language model. Do you have the same observation? Maybe I am doing something wrong

**Comment by chuangchuangtan** [comment]
Llava determines whether it is a base model or a lora model based on the path name, please pay attention to the model save path.

**Comment by Mohan2351999** [comment]
@chuangchuangtan Thanks for the repo, I was able to successfully fine-tune the model on custom dataset for about 20 epochs(trainset around 1.5k samples). I could observe by the end of training the loss was around 0.1 which kind of makes sense. Before training I verified the trainable parameters(i.e. LLM and the projector), the weights for the vision encoder(i.e. vision_tower) are not trainable. After merging the LORA weights, I could observe that LLM and projector weights changed as expected but only a slight shift. But weirdly I also observed that the weights of the vision encoder also slightly shifted. Please find the attachment below. After using the merged weights, the response doesn't align with instruction set. It generates output similar to what LLaVa-Next is doing. Please let me know, what are your observations and any suggestions?

![projector_weight](https://github.com/chuangchuangtan/LLaVA-NeXT-Image-Llama3-Lora/assets/35959745/97f2ae1b-e550-45c7-8b64-17b18fcb2549)
![vision_encoder](https://github.com/chuangchuangtan/LLaVA-NeXT-Image-Llama3-Lora/assets/35959745/aa2e35d3-c471-4e55-ba01-8f5a7c31d199)

Also this how I'm merging the weights:
`python scripts/merge_lora_weights.py --model-path LLaVA-NeXT-Image-Llama3-Lora/checkpoints/llama3-llava-next-8b-lora-20240710_23_41_39 --model-base lmms-lab/llama3-llava-next-8b --save-model-path LLaVA-NeXT-Image-Llama3-Lora/checkpoints/merged_weights`

**Comment by chuangchuangtan** [comment]
It seems to be an issue with llama3-llava-next-8b. We compared llama3-llava-next-8b and openai/clip-vit-large-patch14-336, and found that they are different.
![20240712095200](https://github.com/user-attachments/assets/6cf8a3f3-2751-4339-8917-0d6908272305)

**Comment by chuangchuangtan** [comment]
The [mm_projector](https://github.com/chuangchuangtan/LLaVA-NeXT-Image-Llama3-Lora/blob/5bd7be81b5ab5285c6ae3b7ecd715e5839ba0192/llava/model/llava_arch.py#L90) parameters are set  requires_grad = True.


If you do not want to train mm_projector, you can remove [--mm_projector_lr 2e-5](https://github.com/chuangchuangtan/LLaVA-NeXT-Image-Llama3-Lora/blob/5bd7be81b5ab5285c6ae3b7ecd715e5839ba0192/scripts/llava_llama3/finetune_lora_llava_llama3.sh#L6), and change [code](https://github.com/chuangchuangtan/LLaVA-NeXT-Image-Llama3-Lora/blob/5bd7be81b5ab5285c6ae3b7ecd715e5839ba0192/llava/train/llava_trainer.py#L199):
```python
p for n, p in opt_model.named_parameters() if (n in decay_parameters and 'mm_projector' not in n and p.requires_grad)
```

This setting can avoid [this error](https://github.com/huggingface/peft/issues/137#issue-1599937855). If you have a better way, please let me know.

**Comment by NicoZenith** [comment]
> It seems to be an issue with llama3-llava-next-8b. We compared llama3-llava-next-8b and openai/clip-vit-large-patch14-336, and found that they are different. ![20240712095200](https://private-user-images.githubusercontent.com/19368608/348061968-6cf8a3f3-2751-4339-8917-0d6908272305.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3NzU5MzMsIm5iZiI6MTcyMDc3NTYzMywicGF0aCI6Ii8xOTM2ODYwOC8zNDgwNjE5NjgtNmNmOGEzZjMtMjc1MS00MzM5LTg5MTctMGQ2OTA4MjcyMzA1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzEyVDA5MTM1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTliNDBlMWQ3ZmFlNDE0YzUyNjdkODQ2ODQyNzZjYjRkMzE0NzRiMjgzYjVhMWU1MDBhYjk2NjYxZTVlN2FlYmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.nlXxLWv6BA0Ljz094hJ4qVQyTlIqKzHRzG6c6lXScO8)

In that case it is problematic, because the model you are using for training uses a different vision encoder (openai/clip-vit-large-patch14-336) than the one provided by  llama3-llava-next-8b, so the vision encoder is not aligned with the llama3.

**Comment by chuangchuangtan** [comment]
The config file of llama3-llava-next-8b explicitly states to use openai/clip-vit-large-patch14-336 as the vision tower.

https://huggingface.co/lmms-lab/llama3-llava-next-8b/blob/e7e6a9fd5fd75d44b32987cba51c123338edbede/config.json#L51

**Comment by NicoZenith** [comment]
> The config file of llama3-llava-next-8b explicitly states to use openai/clip-vit-large-patch14-336 as the vision tower.
> 
> https://huggingface.co/lmms-lab/llama3-llava-next-8b/blob/e7e6a9fd5fd75d44b32987cba51c123338edbede/config.json#L51

then what I expect is that the fine-tuning that they did overwrote the weights of the vision encoder, explaining why the values of the weights are changed, while the vision encoder architecture remains the same. What do you think?

**Comment by Mohan2351999** [comment]
@chuangchuangtan, Thanks for the details. The task I'm trying to finetune is predicting a bounding box, and the instruction set looks like this.
![instruction_set](https://github.com/user-attachments/assets/5faee30c-1ca3-47b7-af98-20df99abf7e2)
 The loss is almost at 0.001 at the end of training(Training set has almost 1.5k samples ).
I tried to modify the lora_r and lora_alpha when I set (16, 64) respectively - after training for 20 epochs the output looks very close to the llava-next.
But when I use (64, 16) and (64, 64) - the output doesn't make any sense, with just gives repeated '\n' in the output.
Are you able successfully fine-tune on the instruction set? If so, what is your configurations? and how long did you train for? Also, in the inference script I observed we have parameter "conv_template", do we still leave it as it is if the backbone is "llama3"?
One more thing I don't understand is that, my training loss is almost close to 0.001, which I am assuming its overfitting, but when I try inference on training sample, It can't predict the same output. What is the loss function here? and what are your observations? am I doing something wrong with the inference?

**Comment by NicoZenith** [comment]
> @chuangchuangtan, Thanks for the details. The task I'm trying to finetune is predicting a bounding box, and the instruction set looks like this. ![instruction_set](https://private-user-images.githubusercontent.com/35959745/348287978-5faee30c-1ca3-47b7-af98-20df99abf7e2.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3OTQ1MTMsIm5iZiI6MTcyMDc5NDIxMywicGF0aCI6Ii8zNTk1OTc0NS8zNDgyODc5NzgtNWZhZWUzMGMtMWNhMy00N2I3LWFmOTgtMjBkZjk5YWJmN2UyLlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzEyVDE0MjMzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzZjJkMDFkODVjZDE1NDRmYjM4N2U4Nzg4OTI2YTgyZjZkOTFkNTBjNzIwNGIwNGE1ZDZkNWY3MTQ3NzdiMzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0x8V1UATKerMdKXux7eXWgfQuhdQtrAxTsRWeNbMW1w) The loss is almost at 0.001 at the end of training(Training set has almost 1.5k samples ). I tried to modify the lora_r and lora_alpha when I set (16, 64) respectively - after training for 20 epochs the output looks very close to the llava-next. But when I use (64, 16) and (64, 64) - the output doesn't make any sense, with just gives repeated '\n' in the output. Are you able successfully fine-tune on the instruction set? If so, what is your configurations? and how long did you train for? Also, in the inference script I observed we have parameter "conv_template", do we still leave it as it is if the backbone is "llama3"? One more thing I don't understand is that, my training loss is almost close to 0.001, which I am assuming its overfitting, but when I try inference on training sample, It can't predict the same output. What is the loss function here? and what are your observations? am I doing something wrong with the inference?

This thread here is about fine-tuning llava with llama3 backbone. Your comment is out of scope and should probably addressed to the main repo of LLaVA, in a new issue. The author of the repo is not supposed to address these issues. Thank you :)

**Comment by Mohan2351999** [comment]
Hi @NicoZenith , I was just checking with author, if he has any observations from his finetuning, as this repo's main purpose was finetuning lava-next-llama3. I don't think this is irrelevant, but appreciate if you have insights from your experience fine-tuning on this repo :)

**Comment by NicoZenith** [comment]
@Mohan2351999 yes we can address that in another issue, it's just that we try to solve the biggest issue that vision encoder used for training is not aligned with llama3, that you actually also pointed out :)

**Comment by chuangchuangtan** [comment]
> > The config file of llama3-llava-next-8b explicitly states to use openai/clip-vit-large-patch14-336 as the vision tower.
> > https://huggingface.co/lmms-lab/llama3-llava-next-8b/blob/e7e6a9fd5fd75d44b32987cba51c123338edbede/config.json#L51
> 
> then what I expect is that the fine-tuning that they did overwrote the weights of the vision encoder, explaining why the values of the weights are changed, while the vision encoder architecture remains the same. What do you think?

I think you are right, I will adjust the code to load the vision tower of llava-next.

**Comment by chuangchuangtan** [comment]
> > > The config file of llama3-llava-next-8b explicitly states to use openai/clip-vit-large-patch14-336 as the vision tower.
> > > https://huggingface.co/lmms-lab/llama3-llava-next-8b/blob/e7e6a9fd5fd75d44b32987cba51c123338edbede/config.json#L51
> > 
> > 
> > then what I expect is that the fine-tuning that they did overwrote the weights of the vision encoder, explaining why the values of the weights are changed, while the vision encoder architecture remains the same. What do you think?
> 
> I think you are right, I will adjust the code to load the vision tower of llava-next.

It is expected that modifying [the image encoder code](https://github.com/chuangchuangtan/LLaVA-NeXT-Image-Llama3-Lora/blob/5bd7be81b5ab5285c6ae3b7ecd715e5839ba0192/llava/model/multimodal_encoder/clip_encoder.py#L19) can solve this problem.
```python
        elif getattr(args, 'unfreeze_mm_vision_tower', False) or getattr(args, 'mm_vision_tower_lr', False) :
```

**Comment by chuangchuangtan** [comment]
> @chuangchuangtan, Thanks for the details. The task I'm trying to finetune is predicting a bounding box, and the instruction set looks like this. ![instruction_set](https://private-user-images.githubusercontent.com/35959745/348287978-5faee30c-1ca3-47b7-af98-20df99abf7e2.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA4NDA4MDksIm5iZiI6MTcyMDg0MDUwOSwicGF0aCI6Ii8zNTk1OTc0NS8zNDgyODc5NzgtNWZhZWUzMGMtMWNhMy00N2I3LWFmOTgtMjBkZjk5YWJmN2UyLlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzEzVDAzMTUwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI1NzQwNGU4MjhhYzE4OGNjMjFlMGE2YmQwNDdjOTVhMjJkZjZkNzIxODc3NDY2M2YyZjU0ODI1MWRhZTA2ODAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.FnqzQoTyF_ghnt0VH9UBJ-Cm_kPBrp-nUxSGAKo-y3c) The loss is almost at 0.001 at the end of training(Training set has almost 1.5k samples ). I tried to modify the lora_r and lora_alpha when I set (16, 64) respectively - after training for 20 epochs the output looks very close to the llava-next. But when I use (64, 16) and (64, 64) - the output doesn't make any sense, with just gives repeated '\n' in the output. Are you able successfully fine-tune on the instruction set? If so, what is your configurations? and how long did you train for? Also, in the inference script I observed we have parameter "conv_template", do we still leave it as it is if the backbone is "llama3"? One more thing I don't understand is that, my training loss is almost close to 0.001, which I am assuming its overfitting, but when I try inference on training sample, It can't predict the same output. What is the loss function here? and what are your observations? am I doing something wrong with the inference?

The value 'image' should be the path of image. 
The  parameter  conv_template is necessary. 
Perhaps you can also provide more information to assess the issue.

**Comment by NicoZenith** [comment]
> > > > The config file of llama3-llava-next-8b explicitly states to use openai/clip-vit-large-patch14-336 as the vision tower.
> 
> > > > https://huggingface.co/lmms-lab/llama3-llava-next-8b/blob/e7e6a9fd5fd75d44b32987cba51c123338edbede/config.json#L51
> 
> > > 
> 
> > > 
> 
> > > then what I expect is that the fine-tuning that they did overwrote the weights of the vision encoder, explaining why the values of the weights are changed, while the vision encoder architecture remains the same. What do you think?
> 
> > 
> 
> > I think you are right, I will adjust the code to load the vision tower of llava-next.
> 
> 
> 
> It is expected that modifying [the image encoder code](https://github.com/chuangchuangtan/LLaVA-NeXT-Image-Llama3-Lora/blob/5bd7be81b5ab5285c6ae3b7ecd715e5839ba0192/llava/model/multimodal_encoder/clip_encoder.py#L19) can solve this problem.
> 
> ```python
> 
>         elif getattr(args, 'unfreeze_mm_vision_tower', False) or getattr(args, 'mm_vision_tower_lr', False) :
> 
> ```

I see, but not sure I understand this code, how modifying this code can load the "trained" vision encoder. 

I guess we should create a new path for the vision encoder weights only, extracted from llava-llama-3 weights. What do you think?

**Comment by chuangchuangtan** [comment]
> > > > > The config file of llama3-llava-next-8b explicitly states to use openai/clip-vit-large-patch14-336 as the vision tower.
> > 
> > 
> > > > > https://huggingface.co/lmms-lab/llama3-llava-next-8b/blob/e7e6a9fd5fd75d44b32987cba51c123338edbede/config.json#L51
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > then what I expect is that the fine-tuning that they did overwrote the weights of the vision encoder, explaining why the values of the weights are changed, while the vision encoder architecture remains the same. What do you think?
> > 
> > 
> > > 
> > 
> > 
> > > I think you are right, I will adjust the code to load the vision tower of llava-next.
> > 
> > 
> > It is expected that modifying [the image encoder code](https://github.com/chuangchuangtan/LLaVA-NeXT-Image-Llama3-Lora/blob/5bd7be81b5ab5285c6ae3b7ecd715e5839ba0192/llava/model/multimodal_encoder/clip_encoder.py#L19) can solve this problem.
> > ```python
> >         elif getattr(args, 'unfreeze_mm_vision_tower', False) or getattr(args, 'mm_vision_tower_lr', False) :
> > ```
> 
> I see, but not sure I understand this code, how modifying this code can load the "trained" vision encoder.
> 
> I guess we should create a new path for the vision encoder weights only, extracted from llava-llama-3 weights. What do you think?

There is no need to extract the weights of the vision tower. This approach is feasible. You can check how the vision tower is loaded and verify it by loading the model.
```
from safetensors.torch import load_file as safe_load_file
tensorname='model.vision_tower.vision_tower.vision_model.encoder.layers.9.self_attn.v_proj.weight'
llava_llama3_weight = safe_load_file(‘llama3-llava-next-8b/model-00004-of-00004.safetensors’)[tensorname]
ver_weight               = safe_load_file(‘{Your merge}/model-00004-of-00004.safetensors’)[tensorname]

print((llava_llama3_weight ==ver_weight).min())
```
You can modify this code to validate each tensor.

**Comment by NicoZenith** [comment]
@chuangchuangtan what I mean is that we should now provide a new version of CLIP in the training script, because the base CLIP is not the one that is aligned with the bridger and llama3 via fine-tuning. So I was wondering if we could just extract the vision weights from llama3-llava-next-8b and use it as a vision tower for training.

**Comment by NicoZenith** [comment]
@chuangchuangtan I see you did a new commit yesterday to load llava-llama3 weights. Could you explain me why `mm_vision_tower_lr` will allow to load weights from llava-llama3 instead of openai-clip ?
Thanks a lot! 
Also, do you think we should now also train the vision encoder, as they probably did? Any benefit on training?
